### PR TITLE
Return correct covariance matrix on CovarianceResult

### PR DESCRIPTION
**Pull request by registerrier**
**Description**

This pull request is the result of the discussion in #4149 . It sets the correct covariance matrix on the `CovarianceResult` plus it adds the `Covariance` object on the `FitResult.models` so that all the information should be available inside it.

One should find a proper place to test this.

**Dear reviewer**

### CODE REVIEW

File: gammapy/modeling/fit.py [hunk 1]
```diff
@@ -169,6 +169,10 @@ def run(self, datasets):
 
         covariance_result = self.covariance(datasets=datasets)
 
+        optimize_result.models.covariance = Covariance(
```
Comment (adonath) [inline gammapy/modeling/fit.py:172]: I guess we can actually do the same in `Fit.run()`, such that:
```python3
result = fit.run()
result.models
``` 
Already contains the full covariance...

**Comment by registerrier** [comment]
Normally it should be already doing this.
The following gist shows it works:
https://gist.github.com/registerrier/13194716af259037267c90681c0bea00

Note that, now it is clear that the right object to serialize the `Models` and the `Covariance` is the `FitResult`, not the `Datasets`.

**Comment by adonath** [comment]
Yes, you are right  @registerrier! It was even myself that added this, it seems I just forgot about it.

**Comment by adonath** [comment]
> Note that, now it is clear that the right object to serialize the Models and the Covariance is the FitResult, not the Datasets.

Yes, but it has always been like this. Why would one re-write e.g. all the datasets again, which have not changed during a fit?
Maybe we show this not often enough in tutorials?

**Review by adonath — DISMISSED** [review]
Thanks @registerrier, no further comments from my side.

**Review by AtreyeeS — DISMISSED** [review]
Thanks @registerrier ! This works well.
I think it would be good to modify the fitting tutorial to show `result.models.covariance.plot_correlation`, even though it will not affect the specific example because there is only model.
Also, maybe we can also add a warning or info in `models.correlation` if models is a `DatasetModels)` ?

**Review by adonath — APPROVED** [review]
Thanks @registerrier, looks good to me!

**Pull request merged**